Source/Dafny/Plugins/Compiler.cs

```diff
@@ -140,3 +140,42 @@ public abstract class Compiler {
   public abstract bool RunTargetProgram(string dafnyProgramName, string targetProgramText, string callToMain, string pathsFilename,
     ReadOnlyCollection<string> otherFileNames, object compilationResult, TextWriter outputWriter);
 }
+
+/// <summary>
+/// A placeholder class used to initialize <c>Dafny.Options.O.Compiler</c>.
+/// </summary>
+class PlaceholderCompiler : Compiler {
+  public override IReadOnlySet<string> SupportedExtensions => new HashSet<string>();
+
+  public override string TargetLanguage => "unset";
+  public override string TargetExtension => "unset";
+
+  private Exception UnimplementedExeception() {
+    return new NotImplementedException("DafnyOptions.O.Compiler unset.\n" +
```

Comment:
I see now how an error is thrown now that is much more meaningful, but it's not backward compatible.

What about _extending_ the `CSharpCompiler`  in this placeholder class, and printing a warning message while calling the base classes? At least, things couldn't break, and people would be notified dynamically.
This would be similar to having "deprecated" in public API.

Reply:
OK, that could work.  But why does backward compatibility matter here?  This is a broken usage of the API that trivial (and good!) to fix, no?

Reply:
Sorry I did not answer your question before:

> the API isn't (wasn't?) being used correctly, and it's surfacing that incorrect usage. The right fix is for the client code to call ApplyDefaultOptions().

Previously, there was no need of `ApplyDefaultOptions()` to set up the compiler target to `c#`. It _was_ the default, see here:
https://github.com/dafny-lang/dafny/blob/dc410d680e90429dd3852340021a3a2a37ea974d/Source/Dafny/DafnyOptions.cs#L88

So it became an incorrect usage when the field `CompilationTarget` was replaced with `Compiler`

Reply:
I understand this, but I don't think it's worth fixing beyond this patch (which I already think is a bit too much :/).  If you think it's worth it, I will do it.

Reply:
I think it's worth it. Thanks: -)